sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,708 @@
+                "in X.".format(distinct_clusters, 2),
+                ConvergenceWarning, stacklevel=2)
+
+        return best_centers, best_labels
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Compute bisecting k-means clustering.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+
+            Training instances to cluster.
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
+                                dtype=[np.float64, np.float32],
+                                order='C', copy=self.copy_x,
+                                accept_large_sparse=False)
+
+        self._check_params(X)
+        random_state = check_random_state(self.random_state)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+        self._n_threads = _openmp_effective_n_threads()
+
+        # Validate init array
+        init = self.init
+
+        if hasattr(init, '__array__'):
+            init = check_array(init, dtype=X.dtype, copy=True, order='C')
+            self._validate_center_shape(X, init)
+
+        if self._algorithm == "full":
+            self._kmeans_single = _kmeans_single_lloyd
+            self._check_mkl_vcomp(X, X.shape[0])
+        else:
+            self._kmeans_single = _kmeans_single_elkan
+
+        if self.bisect_strategy == "child_biggest_sse":
+            _bisect_kmeans = self._bisect_by_biggest_child_sse
+
+        elif self.bisect_strategy == "largest_cluster":
+            _bisect_kmeans = self._bisect_largest_cluster
+        else:
+            # "biggest_sse"
+            _bisect_kmeans = self._bisect_by_biggest_sse
+
+        if self.verbose:
+            print("Running Bisecting K-Means with parameters:")
+            print(f"-> number of clusters: {self.n_clusters}")
+            print(f"-> number of centroid initializations: {self.n_init}")
+            print("-> relative tolerance: {:.4e}".format(self.tol))
+            print(f"-> bisect strategy: {self.bisect_strategy} \n")
+
+        _inertia = _inertia_sparse if sp.issparse(X) else _inertia_dense
+
+        # Subtract of mean of X for more accurate distance computations
+        if not sp.issparse(X):
+            X_mean = X.mean(axis=0)
+            X -= X_mean
+
+            if hasattr(init, '__array__'):
+                init -= X_mean
+
+        x_squared_norms = row_norms(X, squared=True)
```

Comment:
If `x_squared_norms` is used only when `self.n_cluster == 1`, then it should be moved inside the `if` clause.
But maybe it could be reused for each bisection, using the same masks as `X` or `sample_weights`.

Reply:
For now I'll move that `self.n_cluster == 1`.
Not really sure if that would be correct with using masks, but I'll need to check definition of `row_norms`

Reply:
It could be reused, but as I tested running each time `row_norms` on passed `data_left` would be faster than one time calculated and pass mask to `_bisect`